// inngest/components/execution/otel/metadataProcessor.js
import { isRecord } from "../../../helpers/types.js";
import { extractAIMetadataFromAttributes } from "./aiExtractor.js";
import { debugPrefix } from "./consts.js";
import Debug from "debug";
import { trace } from "@opentelemetry/api";

//#region src/components/execution/otel/metadataProcessor.ts
const processorDevDebug = Debug(`${debugPrefix}:InngestMetadataSpanProcessor`);
/**
* Resolve the currently-registered global OTel tracer provider, unwrapping
* the `ProxyTracerProvider` wrapper that `trace.getTracerProvider()` returns.
* Returns `undefined` when no provider is registered.
*/
const getGlobalProvider = () => {
	const globalProvider = trace.getTracerProvider();
	if (!globalProvider) return;
	return ("getDelegate" in globalProvider && typeof globalProvider.getDelegate === "function" ? globalProvider.getDelegate() : globalProvider) ?? void 0;
};
/**
* Attempts to add the given span processor to the given OTel provider.
* Returns `true` if the processor was attached, `false` if the provider could
* not be extended.
*
* It handles both OTel SDK v1 (`addSpanProcessor()`) and v2 (internal
* `_spanProcessors` array).
*
* This intentionally duplicates the attach logic inside `extendProvider`
* (`util.ts`) to avoid imports with instrumentation side effects.
*/
const attachToProvider = (provider, processor) => {
	if ("addSpanProcessor" in provider && typeof provider.addSpanProcessor === "function") {
		provider.addSpanProcessor(processor);
		return true;
	}
	const spanProcessors = getInternalSpanProcessors(provider);
	if (spanProcessors) {
		spanProcessors.push(processor);
		return true;
	}
	return false;
};
/**
* Extract the internal span processors array from a BasicTracerProvider.
* Returns the mutable array if accessible, undefined otherwise.
*
* BasicTracerProvider._activeSpanProcessor is a MultiSpanProcessor,
* which holds a _spanProcessors: SpanProcessor[] array.
* Both are TypeScript `private` (not ES #private), so accessible at runtime.
*
* Wrapped in try/catch because this accesses internal OTel fields that may
* change — must never crash the host app.
*/
function getInternalSpanProcessors(provider) {
	if (!isRecord(provider)) return;
	try {
		const active = provider._activeSpanProcessor;
		if (!isRecord(active)) return void 0;
		const arr = active._spanProcessors;
		return Array.isArray(arr) ? arr : void 0;
	} catch {
		return;
	}
}
/**
* Builds the `#spanSinks` key for a span. Span IDs are only guaranteed unique
* within a single trace, so spans are keyed by trace ID + span ID to avoid
* cross-trace collisions.
*/
const spanSinkKey = (traceId, spanId) => `${traceId}:${spanId}`;
/**
* A read-only OTel span processor that is independent of the Extended Traces
* processor (`InngestSpanProcessor`).
*
* It tracks which spans belong to an Inngest step (seeded by
* {@link declareStartingSpan} and passed from parent to child in `onStart`).
*
* When a tracked span ends, it extracts {@link AIMetadata} from the span's
* attributes and pushes it to the span's {@link AIMetadataSink}.
*/
var InngestMetadataSpanProcessor = class {
	/**
	* A map of tracked spans to their sink.
	*
	* We use traceId:spanID as the key, which uniquely identifies each span. See
	* {@link spanSinkKey}
	*
	* The engine seeds the map during {@link declareStartingSpan} with the root
	* span and its sink.
	*
	* During onStart, the processor looks up the span's parent's sink and then
	* records the span as also using that sink. If the parent is not found, then
	* the span is not descended from a root step span, and therefore does not
	* need to have a sink.
	*
	* All spans with the same root span that started the step will share the
	* same sink.
	*/
	#spanSinks = /* @__PURE__ */ new Map();
	/**
	* A registry used to clean up items from `#spanSinks` when spans fall out of
	* reference without ending. Avoids leaking entries (and the engine sink
	* closures they reference) for spans that are never ended and are GC'd.
	*/
	#spanCleanup = new FinalizationRegistry((key) => {
		if (key) this.#spanSinks.delete(key);
	});
	/**
	* Latches once this processor has been attached to a global OTel provider, so
	* {@link attach} can never push it into a provider's processor list twice
	* (which would double-process every span and double-count tokens).
	*/
	#attached = false;
	/**
	* Idempotently attach this processor to the global OTel provider that already
	* exists, so it begins receiving span lifecycle events.
	*/
	attach() {
		if (this.#attached) return;
		const provider = getGlobalProvider();
		if (!provider) return;
		if (attachToProvider(provider, this)) {
			this.#attached = true;
			processorDevDebug("attached to global OTel provider");
		}
	}
	/**
	* Declare the step's root span. Seeds tracking so that the root and all of
	* its descendants share the same AIMetadata sink.
	*/
	declareStartingSpan({ span, traceparent, onAIMetadata }) {
		if (!this.#attached) return;
		if (!traceparent) return processorDevDebug("no traceparent found for span", span.spanContext().spanId, "so skipping it");
		this.trackSpan(span, onAIMetadata);
	}
	/**
	* Mark a span as tracked, recording its step's sink and registering it for
	* cleanup.
	*
	* Read-only: unlike the Extended Traces processor, no attributes
	* are stamped on the span.
	*/
	trackSpan(span, sink) {
		const { traceId, spanId } = span.spanContext();
		const key = spanSinkKey(traceId, spanId);
		this.#spanCleanup.register(span, key, span);
		this.#spanSinks.set(key, sink);
	}
	/**
	* Clean up references to a span that has ended (or been GC'd).
	*/
	cleanupSpan(span) {
		const { traceId, spanId } = span.spanContext();
		this.#spanCleanup.unregister(span);
		this.#spanSinks.delete(spanSinkKey(traceId, spanId));
	}
	/**
	* Track children of spans we already care about, so the whole subtree under a
	* declared root is captured.
	*/
	onStart(span, parentContext) {
		const parentSpanId = trace.getSpanContext(parentContext)?.spanId;
		if (!parentSpanId) return;
		const sink = this.#spanSinks.get(spanSinkKey(span.spanContext().traceId, parentSpanId));
		if (!sink) return;
		this.trackSpan(span, sink);
	}
	/**
	* On end, extract any AI metadata from the span's attributes and push it to
	* its sink, then clean up the span's tracking entry.
	*/
	onEnd(span) {
		const { traceId, spanId } = span.spanContext();
		try {
			const sink = this.#spanSinks.get(spanSinkKey(traceId, spanId));
			if (!sink) return;
			const aiMetadata = extractAIMetadataFromAttributes(span.attributes);
			if (Object.keys(aiMetadata).length === 0) return;
			sink(aiMetadata);
		} finally {
			this.cleanupSpan(span);
		}
	}
	async forceFlush() {}
	async shutdown() {}
};
/**
* The process-wide metadata span processor instance.
*/
const metadataSpanProcessor = new InngestMetadataSpanProcessor();

//#endregion
export { metadataSpanProcessor };
